### PR TITLE
Update | Update SSL certificate error messages

### PR DESCRIPTION
**Pull request by JRahnama**
.Net 7 and greater versions has a new API [X509Certificate2.MatchesHostname](https://learn.microsoft.com/en-us/dotnet/api/system.security.cryptography.x509certificates.x509certificate2.matcheshostname?view=net-8.0) which could be used to determine if a provided hostname exists in certificate SANs or Subject, unfortunately that is not available in lower versions.
This PR updates the .Net 7 and greater versions with the new API and tries to provide more specific error on lower versions.

### CODE REVIEW

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNICommon.cs [hunk 1]
```diff
@@ -180,21 +189,27 @@ internal static bool ValidateSslServerCertificate(string targetServerName, X509C
...
                 {
                     // Fail all other SslPolicy cases besides RemoteCertificateNameMismatch
                     SqlClientEventSource.Log.TrySNITraceEvent(nameof(SNICommon), EventType.ERR, "targetServerName {0}, SslPolicyError {1}, SSL Policy invalidated certificate.", args0: targetServerName, args1: policyErrors);
-                    return false;
+                    throw ADP.SSLCertificateAuthenticationException(Strings.SQL_RemoteCertificateChainErrors);
+                }
+                // the only remaining error would be RemoteCertificateNotAvailable
+                else
+                {
+                    SqlClientEventSource.Log.TrySNITraceEvent(nameof(SNICommon), EventType.INFO, "targetServerName {0}, SSL Policy invalidated certificate.", args0: targetServerName);
```
Comment (Kaur-Parminder) [inline src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNICommon.cs:206]: EventType.ERR for log

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNICommon.cs [hunk 2]
```diff
@@ -5,7 +5,9 @@
 using System;
 using System.Net;
 using System.Net.Security;
+using System.Security.Authentication;
```
Comment (Kaur-Parminder) [inline src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNICommon.cs:8]:
```suggestion
```

Comment (JRahnama) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNICommon.cs:8]: Did you test it with NetStandard2.0 and 2.1? Is it not needed?

Comment (Kaur-Parminder) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNICommon.cs:8]: It show up as unnecessary for NetStandard 2.0 and 2.1 as well.

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNICommon.cs [hunk 3]
```diff
@@ -180,21 +189,27 @@ internal static bool ValidateSslServerCertificate(string targetServerName, X509C
...
                         {
                             SqlClientEventSource.Log.TrySNITraceEvent(nameof(SNICommon), EventType.ERR, "targetServerName {0}, Target Server name does not match Subject in Certificate.", args0: targetServerName);
-                            return false;
+                            throw ADP.SSLCertificateAuthenticationException(Strings.SQL_RemoteCertificateNameMismatch);
                         }
                     }
+#endif
                 }
-                else
+                else if ((policyErrors & SslPolicyErrors.RemoteCertificateChainErrors) != 0)
```
Comment (David-Engel) [inline src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNICommon.cs:197]: Don't use else ifs here. They won't get hit correctly in the event of name plus other policy errors. Evaluate each policy error individually and then name mismatch last, which can have an else block for unknown policy errors.

You can also switch to policyErrors.HasFlag(x) in all these conditions to make the code easier to read.

We can also drill into the chain errors to provide more detail in the error. I'll send you some code suggestions offline.

**Review by David-Engel — COMMENTED** [review]
There is another ValidateSslServerCertificate function below the one being modified that needs similar changes to what you are making (the serverCertificate connection option validation).

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNICommon.cs [hunk 4]
```diff
@@ -180,21 +221,21 @@ internal static bool ValidateSslServerCertificate(string targetServerName, X509C
...
+#endif
                 }
+                // Unknown policy errors
                 else
                 {
-                    // Fail all other SslPolicy cases besides RemoteCertificateNameMismatch
-                    SqlClientEventSource.Log.TrySNITraceEvent(nameof(SNICommon), EventType.ERR, "targetServerName {0}, SslPolicyError {1}, SSL Policy invalidated certificate.", args0: targetServerName, args1: policyErrors);
-                    return false;
+                    SqlClientEventSource.Log.TrySNITraceEvent(nameof(SNICommon), EventType.INFO, "targetServerName {0}, SSL Policy invalidated certificate.", args0: targetServerName);
+                    throw ADP.SSLCertificateAuthenticationException(Strings.SQL_RemoteCertificateNotAvailable);
```
Comment (David-Engel) [inline src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNICommon.cs:233]: We need to use a different error message here.

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNICommon.cs [hunk 5]
```diff
@@ -214,23 +255,63 @@ internal static bool ValidateSslServerCertificate(X509Certificate clientCert, X5
...
+                        throw ADP.SSLCertificateAuthenticationException(Strings.SQL_RemoteCertificateNameMismatch);
                     }
+#endif
                 }
+                // unknown policy errors
                 else
                 {
-                    // Fail all other SslPolicy cases besides RemoteCertificateNameMismatch
                     SqlClientEventSource.Log.TrySNITraceEvent(nameof(SNICommon), EventType.ERR, "certificate subject: {0}, SslPolicyError {1}, SSL Policy invalidated certificate.", args0: clientCert.Subject, args1: policyErrors);
                     return false;
```
Comment (David-Engel) [inline src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNICommon.cs:316]: We should throw the same exception here as we do on line 233.

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNICommon.cs [hunk 6]
```diff
@@ -244,7 +325,7 @@ internal static IPAddress[] GetDnsIpAddresses(string serverName)
             using (TrySNIEventScope.Create(nameof(GetDnsIpAddresses)))
             {
                 SqlClientEventSource.Log.TrySNITraceEvent(nameof(SNICommon), EventType.INFO, "Getting DNS host entries for serverName {0}.", args0: serverName);
-                return Dns.GetHostAddresses(serverName);
+                throw ADP.SSLCertificateAuthenticationException(Strings.SQL_RemoteCertificateNotAvailable);
```
Comment (David-Engel) [inline src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNICommon.cs:328]: Looks like this was accidentally put on the wrong line.

File: src/Microsoft.Data.SqlClient/netcore/src/Resources/Strings.resx [hunk 7]
```diff
@@ -1941,4 +1941,13 @@
   <data name="SQL_SocketDidNotThrow" xml:space="preserve">
     <value>Socket did not throw expected '{0}' with error code '{1}'.</value>
   </data>
-</root>
+  <data name="SQL_RemoteCertificateChainErrors" xml:space="preserve">
+    <value>SSL policy error 'RemoteCertificateChainErrors' detected while validating remote certificate. Chain error message: '{0}'.</value>
```
Comment (David-Engel) [inline src/Microsoft.Data.SqlClient/netcore/src/Resources/Strings.resx:1945]:
```suggestion
    <value>Certificate failed chain validation. Error(s): '{0}'.</value>
```

File: src/Microsoft.Data.SqlClient/netcore/src/Resources/Strings.resx [hunk 8]
```diff
@@ -1941,4 +1941,13 @@
...
+  </data>
+  <data name="SQL_RemoteCertificateNameMismatch" xml:space="preserve">
+    <value>Certificate name mismatch. The provided 'DataSource' or 'HostNameInCertificate' does not match the name in the certificate.</value>
+  </data>
+  <data name="SQL_RemoteCertificateNotAvailable" xml:space="preserve">
+    <value>Certificate not available while validating the server certificate.</value>
```
Comment (David-Engel) [inline src/Microsoft.Data.SqlClient/netcore/src/Resources/Strings.resx:1951]: Slight tweak to make the message match the others.
```suggestion
    <value>Certificate not available while validating the certificate.</value>
```

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNICommon.cs (hunk 6 above)
Comment (JRahnama) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNICommon.cs:328]: oops, this is GitHub Copilot insertion. I did not pay attention to this...

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNICommon.cs [hunk 9]
```diff
@@ -146,31 +148,70 @@ internal static bool ValidateSslServerCertificate(string targetServerName, X509C
...
+                    bool chainIsValid = chain.Build(cert2);
+                    if (!chainIsValid)
+                    {
+                        foreach (X509ChainStatus chainStatus in chain.ChainStatus)
+                        {
+                            if (chainStatusInformation.Length == 0)
+                            {
+                                chainStatusInformation.Append(Environment.NewLine);
+                            }
+                            chainStatusInformation.Append(chainStatus.StatusInformation);
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNICommon.cs:169]: Can you include the `Status` enum in the `chainStatusInformation` as well?
```suggestion
                            chainStatusInformation.Append(chainStatus.StatusInformation);
```

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNICommon.cs [hunk 10]
```diff
@@ -180,21 +221,22 @@ internal static bool ValidateSslServerCertificate(string targetServerName, X509C
...
                 }
+                // Unknown policy errors
+                // It should not get here as all the policy errors are handled above, just to cover some unknown cases.
                 else
                 {
-                    // Fail all other SslPolicy cases besides RemoteCertificateNameMismatch
-                    SqlClientEventSource.Log.TrySNITraceEvent(nameof(SNICommon), EventType.ERR, "targetServerName {0}, SslPolicyError {1}, SSL Policy invalidated certificate.", args0: targetServerName, args1: policyErrors);
-                    return false;
+                    SqlClientEventSource.Log.TrySNITraceEvent(nameof(SNICommon), EventType.ERR, "targetServerName {0}, SSL Policy invalidated certificate.", args0: targetServerName);
+                    throw ADP.SSLCertificateAuthenticationException(Strings.SQL_RemoteCertificateUnknownError);
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNICommon.cs:234]: Can you include the `policyErrors` enum in this exception?

**Review by DavoudEshtehari — COMMENTED** [review]
I'd suggest collecting all the possible exceptions and throwing an exception at the end.

**Review by David-Engel — APPROVED** [review]

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/ConnectionTestWithSSLCert/CertificateTest.cs [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+        public void OpeningConnectionWitHNICTest()
+        {
+            // Mandatory
+            SqlConnectionStringBuilder builder = new(DataTestUtility.TCPConnectionString)
+            {
+                // 127.0.0.1 most of the cases does not cause any Remote certificate validation error depinding on name resolution on the machine
+                //  It mostly returns SslPolicyErrors.None
+                DataSource = IPV4,
+                Encrypt = SqlConnectionEncryptOption.Mandatory,
+                HostNameInCertificate = "localhost"
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/ConnectionTestWithSSLCert/CertificateTest.cs:76]: Can you check whether this works without specifying this in Native SNI?
Also do run all tests with both Managed and Native SNI whenever possible, and behavior should be identical.

File: src/Microsoft.Data.SqlClient/tests/ManualTests/DataCommon/DataTestUtility.cs [hunk 12]
```diff
@@ -85,6 +86,29 @@ public static class DataTestUtility
...
+                return s_sQLServerVersion;
+            }
+        }
+
+        // Is TDS8 supported
+        public static bool IsTDS8Supported
+        {
+            get
+            {
+                s_isTDS8Supported = GetSQLServerStatusOnTDS8(TCPConnectionString);
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/DataCommon/DataTestUtility.cs:107]: I see you've already validated the `TCPConnectionString` in the tests, but as a stand alone utility it's good to validate these properties as well and throw exception.

File: src/Microsoft.Data.SqlClient/tests/ManualTests/DataCommon/DataTestUtility.cs [hunk 13]
```diff
@@ -235,8 +259,44 @@ private static string GenerateAccessToken(string authorityURL, string aADAuthUse
             return result.AccessToken;
         });
 
+        public static bool IsManagedSNI => UseManagedSNIOnWindows;
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/DataCommon/DataTestUtility.cs:262]: I'd rather using the `UseManagedSNIOnWindows` and avoid adding a new field.

File: src/Microsoft.Data.SqlClient/tests/ManualTests/DataCommon/DataTestUtility.cs [hunk 14]
```diff
@@ -235,8 +259,44 @@ private static string GenerateAccessToken(string authorityURL, string aADAuthUse
...
+            try
+            {
+                SqlConnection conn = new(builder.ConnectionString);
+                conn.Open();
+                isTDS8Supported = true;
+            }
+            catch (SqlException)
+            {
+
+            }
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/DataCommon/DataTestUtility.cs:296]:
```suggestion
            SqlConnection conn = new(builder.ConnectionString);
            conn.Open();
            isTDS8Supported = true;
```

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/ConnectionTestWithSSLCert/CertificateTest.cs [hunk 15]
```diff
@@ -0,0 +1,259 @@
...
+        }
+        #endregion
+
+        public CertificateTest()
+        {
+            CreateValidCertificate(s_fullPathToPowershellScript);
+            _thumbprint = Environment.GetEnvironmentVariable(ThumbPrintEnvName, EnvironmentVariableTarget.Machine);
+        }
+
+        [ConditionalFact(typeof(DataTestUtility), nameof(DataTestUtility.AreConnStringsSetup), nameof(DataTestUtility.IsAdmin), nameof(DataTestUtility.IsNotAzureServer))]
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/ConnectionTestWithSSLCert/CertificateTest.cs:57]: Checking the admin permission would silently skipped the test. I'd check it inside the test and throw exception instead.
Same note for all the consumptions.

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/ConnectionTestWithSSLCert/GenerateSelfSignedCertificate.ps1 [hunk 16]
```diff
@@ -0,0 +1,123 @@
+# Script: Invoke-SqlServerCertificateCommand
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/ConnectionTestWithSSLCert/GenerateSelfSignedCertificate.ps1:1]:
```suggestion
# Licensed to the .NET Foundation under one or more agreements.
# The .NET Foundation licenses this file to you under the MIT license.
# See the LICENSE file in the project root for more information.
# Script: Invoke-SqlServerCertificateCommand
```

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/ConnectionTestWithSSLCert/CertificateTest.cs (hunk 15 above)
Comment (JRahnama) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/ConnectionTestWithSSLCert/CertificateTest.cs:57]: we want to skip the test if it is not running as admin. not sure if throwing exception would help here.

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/ConnectionTestWithSSLCert/CertificateTest.cs [hunk 17]
```diff
@@ -0,0 +1,259 @@
+﻿using System;
```
Comment (David-Engel) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/ConnectionTestWithSSLCert/CertificateTest.cs:1]:
```suggestion
# Licensed to the .NET Foundation under one or more agreements.
# The .NET Foundation licenses this file to you under the MIT license.
# See the LICENSE file in the project root for more information.
﻿using System;
```

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/ConnectionTestWithSSLCert/CertificateTest.cs [hunk 18]
```diff
@@ -0,0 +1,268 @@
...
+        {
+            if (DataTestUtility.IsAdmin)
+            {
+                CreateValidCertificate(s_fullPathToPowershellScript);
+                _thumbprint = Environment.GetEnvironmentVariable(ThumbPrintEnvName, EnvironmentVariableTarget.Machine);
+            }
+            else
+            {
+                Assert.False(true, "CertificateTest class needs to be run in Admin mode.");
+            }
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/ConnectionTestWithSSLCert/CertificateTest.cs:63]:
```suggestion
            Assert.True(DataTestUtility.IsAdmin, "CertificateTest class needs to be run in Admin mode.");

            CreateValidCertificate(s_fullPathToPowershellScript);
            _thumbprint = Environment.GetEnvironmentVariable(ThumbPrintEnvName, EnvironmentVariableTarget.Machine);
```

**Review by DavoudEshtehari — APPROVED** [review]

**Pull request merged**